include/xtensor/views/xslice.hpp

```diff
@@ -564,78 +543,48 @@ namespace xt
-            xrange<std::ptrdiff_t>>
-        get(std::size_t size) const
-        {
-            return xrange<std::ptrdiff_t>(0, normalize(m_stop, size));
-        }
-
-        template <class MI = A, class MA = B, class STEP = C>
-        inline std::enable_if_t<
-            !xtl::is_integral<MI>::value && !xtl::is_integral<MA>::value && !xtl::is_integral<STEP>::value,
-            xall<std::ptrdiff_t>>
-        get(std::size_t size) const
+        auto get(std::size_t size) const
         {
-            return xall<std::ptrdiff_t>(static_cast<std::ptrdiff_t>(size));
+            if constexpr (xtl::is_integral<MI>::value && xtl::is_integral<MA>::value && xtl::is_integral<STEP>::value)
+            {
+                return get_stepped_range(m_start, m_stop, m_step, size);
+            }
+            else if constexpr (!xtl::is_integral<MI>::value && xtl::is_integral<MA>::value && xtl::is_integral<STEP>::value)
+            {
+                return get_stepped_range(
+                    m_step > 0 ? 0 : static_cast<std::ptrdiff_t>(size) - 1,
+                    m_stop,
+                    m_step,
+                    size
+                );
+            }
+            else if constexpr (xtl::is_integral<MI>::value && !xtl::is_integral<MA>::value && xtl::is_integral<STEP>::value)
+            {
+                auto sz = static_cast<std::ptrdiff_t>(size);
+                return get_stepped_range(m_start, m_step > 0 ? sz : -(sz + 1), m_step, size);
+            }
+            else if constexpr (xtl::is_integral<MI>::value && xtl::is_integral<MA>::value && !xtl::is_integral<STEP>::value)
+            {
+                return xrange<std::ptrdiff_t>(normalize(m_start, size), normalize(m_stop, size));
+            }
+            else if constexpr (!xtl::is_integral<MI>::value && !xtl::is_integral<MA>::value && xtl::is_integral<STEP>::value)
+            {
+                std::ptrdiff_t start = m_step >= 0 ? 0 : static_cast<std::ptrdiff_t>(size) - 1;
+                std::ptrdiff_t stop = m_step >= 0 ? static_cast<std::ptrdiff_t>(size) : -1;
+                return xstepped_range<std::ptrdiff_t>(start, stop, m_step);
+            }
+            else if constexpr (xtl::is_integral<MI>::value && !xtl::is_integral<MA>::value && !xtl::is_integral<STEP>::value)
+            {
+                return xrange<std::ptrdiff_t>(normalize(m_start, size), static_cast<std::ptrdiff_t>(size));
+            }
+            else if constexpr (!xtl::is_integral<MI>::value && xtl::is_integral<MA>::value && !xtl::is_integral<STEP>::value)
+            {
+                return xrange<std::ptrdiff_t>(0, normalize(m_stop, size));
+            }
+            else if constexpr (!xtl::is_integral<MI>::value && !xtl::is_integral<MA>::value && !xtl::is_integral<STEP>::value)
+            {
+                return xall<std::ptrdiff_t>(static_cast<std::ptrdiff_t>(size));
+            }
```

Comment:
Note for later: It might be worth having more expressive metafunctions wrapping these conditions (and somehow categorizing the slice types), especially if they are used elsewhere. This should be done in a dedicated PR though.